janusgraph-core/src/main/java/org/janusgraph/graphdb/types/indextype/CompositeIndexTypeWrapper.java

```diff
@@ -58,7 +58,7 @@ public SchemaStatus getStatus() {
         return base.getStatus();
     }
 
-    IndexField[] fields = null;
+    volatile IndexField[] fields = null;
```

Comment:
I can't say if `volatile` is OK here or not. Again, I'm not much familiar with this part of the code but my thoughts are the next:
If we check `getFieldKeys()` method implementation we could see that at the beginning the author explicitly copies the reference from class variable to method level variable to not lose the reference. I.e. the author expects that the code could be executed concurrently but instead of synchronizing the code, the author prefers to explicitly compute this code multiple times if such situation acquires. I.e. looks like the author thinks that race condition will happen rarely here but if it ever happen it is OK to compute the same code several times.
Now about `volatile`. It makes sense of course to compute this code only once and then to reuse the variable by other threads but `volatile` will force us to always read and write this variable directly to RAM and never to CPU cache which could potentially affect performance.
I believe the author didn't put `volatile` here on purpose because most likely it will rarely happen that all threads which go here read stale variable from CPU cache. It can only happen if there are already multiple existing threads and all of them cached `null` in CPU and then they execute this code sequentially and CPU don't clear that variable from cache. Sounds very unlikely but even if that happens, we are back to synchronization problem. As explained above, it looks like the author didn't want to apply order guarantee to this part of the code because it's okay to compute this code multiple times and the cashed result can be either `null` or the same as in other threads (i.e. could be different instance of the array but the result in that array is the same as in other threads CPU cache).
Again, the above are just my thoughts. I can't say for sure if we need `volatile` here (and below) or not.